pkg/ccl/sqlproxyccl/connector.go

```diff
@@ -0,0 +1,389 @@
+	}
+	return crdbConn, false, nil
+}
+
+// openClusterConnInternal returns a connection to the tenant cluster associated
+// with the connector. Once a connection has been established, the pgwire
+// startup message will be relayed to the server. Returned errors may be marked
+// as a lookup or dial error.
+func (c *connector) openClusterConnInternal(ctx context.Context) (net.Conn, error) {
+	if c.testingKnobs.openClusterConnInternal != nil {
+		return c.testingKnobs.openClusterConnInternal(ctx)
+	}
+
+	// Repeatedly try to make a connection until context is canceled, or until
+	// we get a non-retriable error. This is preferable to terminating client
+	// connections, because in most cases those connections will simply be
+	// retried, further increasing load on the system.
+	retryOpts := retry.Options{
+		InitialBackoff: 10 * time.Millisecond,
+		MaxBackoff:     5 * time.Second,
+	}
+
+	outgoingAddressErr := log.Every(time.Minute)
+	backendDialErr := log.Every(time.Minute)
+	reportFailureErr := log.Every(time.Minute)
+	var outgoingAddressErrs, codeBackendDownErrs, reportFailureErrs int
+	var crdbConn net.Conn
+	var outgoingAddr string
+	var err error
+
+	for r := retry.StartWithCtx(ctx, retryOpts); r.Next(); {
+		// Retrieve a SQL pod address to connect to.
+		outgoingAddr, err = c.lookupAddr(ctx)
+		if err != nil {
+			if isRetriableConnectorError(err) {
+				outgoingAddressErrs++
+				if outgoingAddressErr.ShouldLog() {
+					log.Ops.Errorf(ctx, "lookup address (%d errors skipped): %v",
+						outgoingAddressErrs, err)
+					outgoingAddressErrs = 0
+				}
+				continue
+			}
+			return nil, err
+		}
+		// Make a connection to the SQL pod.
+		crdbConn, err = c.dialOutgoingAddr(outgoingAddr)
+		if err != nil {
+			if isRetriableConnectorError(err) {
+				codeBackendDownErrs++
+				if backendDialErr.ShouldLog() {
+					log.Ops.Errorf(ctx, "dial outgoing address (%d errors skipped): %v",
+						codeBackendDownErrs, err)
+					codeBackendDownErrs = 0
+				}
+
+				// Report the failure to the directory so that it can refresh
+				// any stale information that may have caused the problem.
+				if c.Directory != nil {
+					if err = reportFailureToDirectory(
+						ctx, c.TenantID, outgoingAddr, c.Directory,
+					); err != nil {
+						reportFailureErrs++
+						if reportFailureErr.ShouldLog() {
+							log.Ops.Errorf(ctx,
+								"report failure (%d errors skipped): %v",
+								reportFailureErrs,
+								err,
+							)
+							reportFailureErrs = 0
+						}
+					}
+				}
+				continue
+			}
+			return nil, err
+		}
+		return crdbConn, nil
+	}
+
+	// Since the retry loop above retries infinitely, the only possibility
+	// where we will exit the loop is when context is cancelled.
+	if errors.Is(err, context.Canceled) {
+		return nil, err
+	}
+	// Loop exited at boundary, so mark previous error with cancellation.
+	if ctxErr := ctx.Err(); err != nil && ctxErr != nil {
+		return nil, errors.Mark(err, ctxErr)
+	}
+	panic("unreachable")
```

Comment:
nit: avoid panics in the data plane code, even if it is unreachable. Log the error as unexpected and return. Alternatively, you could restructure the loop so that this is impossible. E.g. make the only break condition in the loop:
```
if ctx.Err() != nil {
  return nil, ctx.Err()
}
```